flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala

```diff
@@ -411,6 +411,23 @@ class TaskManager(
             log.debug(s"Cannot find task to fail for execution ${executionID})")
           }
 
+        // stops a task
+        case StopTask(executionID) =>
+          val task = runningTasks.get(executionID)
+          if (task != null) {
+            try {
+              task.stopExecution()
```

Comment:
I agree with @mjsax that the cancel should be non-blocking. The method seems to be badly named, it should maybe have been named `stop()` from the beginning. Then it would be more obvious that is is just used for shutting down the source, no matter what the reason might be.

Unfortunately `close()` is also taken as part of `RichFunction`.

I think we have two options, keep `cancel` and use it for stopping, no matter what the reason is. Or rename it to `stop()` and use it in the same way to make it clearer what it does.